### PR TITLE
Implement recording to MP4 or WebM

### PR DESCRIPTION
**Pull request by sdroege**
CC @GuillaumeGomez 

Next I'll want to clean up some more things (e.g. move functions into the `App` impl block) and add better error reporting (error dialogs). But this part can be reviewed already.

Do you understand what the GStreamer code is doing based on the comments or do the comments need improvement?

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -261,6 +311,43 @@ fn create_pipeline(app: &App) -> Result<(gst::Pipeline, gtk::Widget), Box<dyn er
                 inner.error = Some(Box::new(err.get_error()));
                 inner.application.quit();
             }
+            MessageView::Element(msg) => {
+                // Catch the EOS messages from our filesink. Because the other sink,
```
Comment (GuillaumeGomez) [inline src/main.rs]: What is EOS?

File: src/main.rs [hunk 2]
```diff
@@ -573,6 +722,131 @@ fn on_snapshot_button_clicked(
...
+        let tee = pipeline.get_by_name("tee").unwrap();
+        let srcpad = tee.get_request_pad("src_%u").unwrap();
+        let sinkpad = bin.get_static_pad("sink").unwrap();
+
+        // If linking fails, we just undo what we did above
+        if let Err(err) = srcpad.link(&sinkpad).into_result() {
+            // TODO error dialogs
+            eprintln!("Failed to link recording bin: {}", err);
+            pipeline.remove(&bin).unwrap();
+            let _ = bin.set_state(gst::State::Null);
```
Comment (GuillaumeGomez) [inline src/main.rs]: Is it returning an error or something?

**Comment by GuillaumeGomez** [comment]
I commented where I had small issues. However, I'd really like if you used `expect` instead of `unwrap`. :)

File: src/main.rs (hunk 1 above)
Comment (sdroege) [reply src/main.rs]: Ah I wrote it out elsewhere but forgot to do so here. Thanks :) End Of Stream

File: src/main.rs (hunk 2 above)
Comment (sdroege) [reply src/main.rs]: Yeah, but here we're in an error path anyway so any further error should be just ignored. Will add a comment, thanks!

**Comment by sdroege** [comment]
> However, I'd really like if you used `expect` instead of `unwrap`. :)

All replaced locally. Ok, let's hope I can finish the changes at least tomorrow morning :)

**Comment by GuillaumeGomez** [comment]
Thanks a lot!

**Comment by sdroege** [comment]
@GuillaumeGomez how's this?

**Comment by GuillaumeGomez** [comment]
Yep, all good! :+1:

**Pull request merged**